src/solvers/smt2/smt2_parser.cpp

```diff
@@ -950,7 +950,22 @@ void smt2_parsert::setup_expressions()
 
   expressions["distinct"] = [this] {
     // pair-wise different constraint, multi-ary
-    return multi_ary("distinct", operands());
+    auto op = operands();
+    if(op.size() % 2 != 0)
+      throw error("distinct needs even number of operands");
+    if(op.size() == 2)
+      return binary_predicate(ID_notequal, op);
+    else
+    {
+      std::vector<exprt> pairwise_constraints;
+      for(std::size_t i = 0; i < op.size();)
```

Comment:
I don't think this is the semantics of SMT-LIB `distinct`.  It is described here ( http://smtlib.cs.uiowa.edu/theories-Core.shtml ) as:

```
(par (A) (distinct A A Bool :pairwise))
```

And `:pairwise` is described on Page 34 of http://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2017-07-18.pdf as:

```
[pairwise annotation] (f   t1···tn) is syntactic sugar (recursively) for (and (f   t1 t2) ··· (f   t1 tn) (f   t2···tn)).
```